### PR TITLE
7955 libshare needs to initialize only those datasets being modified by the consumer

### PR DESCRIPTION
**Pull request by prakashsurya**
Reviewed by: Steve Gonczi <steve.gonczi@delphix.com>
Reviewed by: Sebastien Roy <sebastien.roy@delphix.com>
Reviewed by: Matthew Ahrens <mahrens@delphix.com>
Reviewed by: George Wilson <george.wilson@delphix.com>
Reviewed by: Pavel Zakharov <pavel.zakharov@delphix.com>

Libshare currently initializes all available filesystems when doing any
libshare operation. This requires iterating through all the filesystem
multiple times, which is a huge performance problem for sharing and
unsharing operations.

This patch modifies the initialization of libshare in two more modes
beyond what had happened before. In one, a pathname is passed in, and
that one path is intiialized if possible. In another mode of operation,
an array of zfs dataset handles is passed into libshare to be
initialized. In any case, these initialization operations should speed
up most uses of libshare.

The following tests were run on a system with ~8300 ZFS filesystems on
it, using time(1) to measure the performance of each command without
this patch:

    $ sudo zfs create -o sharenfs=on rpool/test
    $ time sudo zfs destroy rpool/test
    real    0m28.957s
    user    0m7.494s
    sys     0m20.969s

    $ time sudo zfs create -o sharenfs=on rpool/test
    real    0m29.216s
    user    0m7.517s
    sys     0m20.579s

    $ time sudo zfs destroy rpool/test
    real    0m28.527s
    user    0m7.435s
    sys     0m20.688s

And then same commands were run on the same system, but using a version
of libshare with this patch applied:

    $ sudo zfs create -o sharenfs=on rpool/test
    $ time sudo zfs destroy rpool/test
    real    0m0.624s
    user    0m0.067s
    sys     0m0.255s

    $ time sudo zfs create -o sharenfs=on rpool/test
    real    0m0.797s
    user    0m0.020s
    sys     0m0.183s

    $ time sudo zfs destroy rpool/test
    real    0m0.627s
    user    0m0.072s
    sys     0m0.270s

That's about a 97% decrease in "wall clock" latency for these commands.

Upstream bugs: DLPX-47705, DLPX-47160, DLPX-48226, DLPX-50300

### CODE REVIEW

**Comment by prakashsurya** [comment]
@zettabot go

**Comment by prakashsurya** [comment]
@zettabot go

File: usr/src/lib/libshare/common/libshare.h [hunk 1]
```diff
@@ -22,6 +22,8 @@
 /*
  * Copyright (c) 2006, 2010, Oracle and/or its affiliates. All rights reserved.
  * Copyright 2015 Nexenta Systems, Inc.  All rights reserved.
+ * Copyright 2013 Nexenta Systems, Inc.  All rights reserved.
```
Comment (yuripv) [inline usr/src/lib/libshare/common/libshare.h]: looks like mismerge.

File: usr/src/test/zfs-tests/tests/functional/inheritance/inherit_001_pos.ksh [hunk 2]
```diff
@@ -331,7 +331,7 @@ function scan_state { #state-file
 				if [[ $op == "-" ]]; then
 					log_note "No operation specified"
 				else
-					export __ZFS_POOL_RESTRICT="$TESTPOOL"
+					export __ZFS_POOL_RESTRICT="TESTPOOL"
```
Comment (yuripv) [inline usr/src/test/zfs-tests/tests/functional/inheritance/inherit_001_pos.ksh]: Why?

Comment (prakashsurya) [inline usr/src/test/zfs-tests/tests/functional/inheritance/inherit_001_pos.ksh]: I wasn't the original author, and haven't had time to double check this specific change. I think it was added due to failures seen in this specific test case, but I see that we already ignore failures of this test in our testing (see https://github.com/openzfs/openzfs-ci/blob/master/jenkins/sh/run-zfstest/zfstest-report.py#L51), so maybe this should be removed from the diff?

@jwk404 do you remember why this line was changed?

**Comment by prakashsurya** [comment]
@zettabot go

File: usr/src/test/zfs-tests/tests/functional/inheritance/inherit_001_pos.ksh (hunk 2 above)
Comment (yuripv) [inline usr/src/test/zfs-tests/tests/functional/inheritance/inherit_001_pos.ksh]: got it (after reading the test case itself).

**Comment by prakashsurya** [comment]
@zettabot go

**Comment by prakashsurya** [comment]
@yuripv do you have any more questions about this one? can I move forward with the RTI? and if so, should I count you as a reviewer?

**Comment by yuripv** [comment]
No more questions, looks good.

**Comment by rmustacc** [comment]
@prakashsurya, @djhoffma we've had a few folks stumble across some issues related to this change. @jjelinek was looking and concluded that if libshare isn't installed or is missing, it basically means that zfs mount always fails, even when it used to work before. This happens for datasets regardless if any share properties are present.

Given the design of commands to not assume that libshare is present, it seems like this should be changed to drive on in the face of it, even if that means that sharing isn't occurring. Thoughts?

**Comment by prakashsurya** [comment]
@rmustacc I didn't realize that was possible (to not have libshare installed, while still using the zfs commands). I think your suggestion makes sense; since it's possible to not have libshare installed, we should probably avoid the failure, and handle that case more gracefully (maintaining the prior behavior).

**Comment by rmustacc** [comment]
@prakashsurya Yeah. ZFS goes to great pains to do the entire dlopen trick to try and deal with that. Otherwise you'd just link against libshare and call it a day (I'd assume). We have a change for that which we're testing locally at the moment. Effectively it's making the init return SA_OK in this case, since the rest of ZFS is built around assuming that libshare is missing.

**Comment by jjelinek** [comment]
This is the diff I've tested within an lx-branded zone with a delegated dataset. Running "/native/usr/sbin/zfs mount -a" now works.

--- a/usr/src/lib/libzfs/common/libzfs_mount.c
+++ b/usr/src/lib/libzfs/common/libzfs_mount.c
@@ -662,8 +663,9 @@ _zfs_init_libshare(void)
 static int
 zfs_init_libshare_impl(libzfs_handle_t *zhandle, int service, void *arg)
 {
+       /* libshare is either not installed or we're in a branded zone */
        if (_sa_init == NULL)
-               return (SA_CONFIG_ERR);
+               return (SA_OK);
 
        /*

**Comment by prakashsurya** [comment]
Sorry for the regression... @jjelinek would you like to open a PR with the fix? or, alternatively, I can do that and do the legwork to get the fix integrated to illumos (assuming I can lean on you to test it in your libshare-less configuration).

**Comment by rmustacc** [comment]
@prakashsurya we'll end up doing some additional testing and sync in illumos-joyent first before we upstream it. I suspect that due to the way IPS packaging works, we'll only see this in some of the odder zone brands that we have.

**Comment by prakashsurya** [comment]
OK, sounds good. Let me know if there's anything I can do to help get the fix integrated (reviews, testing, etc). Sorry for the hassle.

**Pull request closed without merging**